### PR TITLE
[WIP] Use interpolated strings

### PR DESCRIPTION
**Pull request by CarloToso**
# PR Summary

Use interpolated strings with `String.Format()`, `StringBuilder.AppendFormat()`, `String.Create()`

## PR Context
Proposed by @iSazonov 

## PR Checklist

- [ ] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [ ] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs [hunk 1]
```diff
@@ -1031,11 +1031,8 @@ internal FileRedirection(RedirectionStream from, bool appending, string file)
 
         public override string ToString()
         {
-            return string.Format(CultureInfo.InvariantCulture, "{0}> {1}",
-                                 FromStream == RedirectionStream.All
-                                     ? "*"
-                                     : ((int)FromStream).ToString(CultureInfo.InvariantCulture),
-                                 File);
+            return string.Format(CultureInfo.InvariantCulture, $"{0}> {File}",
+                                 FromStream == RedirectionStream.All ? "*" : ((int)FromStream).ToString(CultureInfo.InvariantCulture));
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs:1035]: The code can be insterted in {0}.
(In C# 11 even multilines are supported https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-11#newlines-in-string-interpolations but we should switch from C# 10 to 11 in our msbuild files.)

File: src/System.Management.Automation/utils/tracing/PSEtwLog.cs [hunk 2]
```diff
@@ -245,7 +245,7 @@ internal static void LogAnalyticWarning(PSEventId id, PSOpcode opcode, PSTask ta
             if (provider.IsEnabled(PSLevel.Verbose, keyword))
             {
                 string payLoadData = BitConverter.ToString(fragmentData.blob, fragmentData.offset, fragmentData.length);
-                payLoadData = string.Format(CultureInfo.InvariantCulture, "0x{0}", payLoadData.Replace("-", string.Empty));
+                payLoadData = string.Format(CultureInfo.InvariantCulture, $"0x{payLoadData.Replace("-", string.Empty)}");
```
Comment (iSazonov) [inline src/System.Management.Automation/utils/tracing/PSEtwLog.cs:248]: Perhaps follow works (or will in .Net 8, or C# 11)
```c#
$"0x{payLoadData.AsSpan().Replace("-", string.Empty)}");
```

File: src/Microsoft.Management.Infrastructure.CimCmdlets/CimSessionOperations.cs [hunk 3]
```diff
@@ -951,7 +951,7 @@ internal void AddSessionToCache(CimSession cimSession, XOperationContextBase con
             CimTestCimSessionContext testCimSessionContext = context as CimTestCimSessionContext;
             uint sessionId = this.sessionState.GenerateSessionId();
             string originalSessionName = testCimSessionContext.CimSessionWrapper.Name;
-            string sessionName = originalSessionName ?? string.Format(CultureInfo.CurrentUICulture, @"{0}{1}", CimSessionState.CimSessionClassName, sessionId);
```
Comment (iSazonov) [inline src/Microsoft.Management.Infrastructure.CimCmdlets/CimSessionOperations.cs:954]: Really we should replace `string.Format` with [`string.Create`](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/String.cs,316)

Comment (CarloToso) [reply src/Microsoft.Management.Infrastructure.CimCmdlets/CimSessionOperations.cs:954]: Follow up PR?

Comment (iSazonov) [reply src/Microsoft.Management.Infrastructure.CimCmdlets/CimSessionOperations.cs:954]: No, current changes haven't value. Idea is to use _new_ API to reduce allocations. There is no `string.Format` with interpalated string handler. I guess I misled you. Please use `string.Create`.
(You can find all new APIs with interpolated handlers starting with https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/InterpolatedStringHandlerAttribute.cs,382762ca0bf5bcac,references)

Comment (iSazonov) [reply src/Microsoft.Management.Infrastructure.CimCmdlets/CimSessionOperations.cs:954]: Also please don't change our trace.WriteLine() and other such callsites - this requires changes in the APIs themselves, which is not always easy. See example #18246.

File: src/Microsoft.PowerShell.Commands.Diagnostics/ImportCounterCommand.cs [hunk 4]
```diff
@@ -240,7 +240,7 @@ protected override void EndProcessing()
                     break;
 
                 default:
-                    Debug.Assert(false, string.Format(CultureInfo.InvariantCulture, "Invalid parameter set name: {0}", ParameterSetName));
+                    Debug.Assert(false, string.Format(CultureInfo.InvariantCulture, $"Invalid parameter set name: {ParameterSetName}"));
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Diagnostics/ImportCounterCommand.cs:243]: Debug.Assert supports interpolated string handler and we can use interpolated string directly [Debug.Assert](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Debug.cs,89)
```suggestion
                    Debug.Assert(false, $"Invalid parameter set name: {ParameterSetName}");
```

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs [hunk 5]
```diff
@@ -1522,7 +1519,7 @@ private static void WriteConsoleOutputCJK(ConsoleHandle consoleHandle, Coordinat
                             // to write is larger than bufferLimit. In that case, the algorithm writes one row
                             // at a time => bufferSize.Y == 1. Then, we can safely leave bufferSize.Y unchanged
                             // to retry with a smaller bufferSize.X.
-                            Dbg.Assert(bufferSize.Y == 1, string.Format(CultureInfo.InvariantCulture, "bufferSize.Y should be 1, but is {0}", bufferSize.Y));
+                            Dbg.Assert(bufferSize.Y == 1, string.Format(CultureInfo.InvariantCulture, $"bufferSize.Y should be 1, but is {bufferSize.Y}"));
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs:1522]: `Dbg.Assert` is our API. In **follow** PR we could modernize it to support interpolated string handler and then remove `string.Format` from such callsites..

File: src/Microsoft.Management.Infrastructure.CimCmdlets/CimSessionOperations.cs [hunk 6]
```diff
@@ -951,7 +951,7 @@ internal void AddSessionToCache(CimSession cimSession, XOperationContextBase con
...
+            string sessionName = originalSessionName ?? string.Create(CultureInfo.CurrentUICulture, $@"{CimSessionState.CimSessionClassName}{sessionId}");
```
Comment (iSazonov) [inline src/Microsoft.Management.Infrastructure.CimCmdlets/CimSessionOperations.cs:954]: We can remove  verbatim string literal `@`. Below I see the same too.

File: src/Microsoft.Management.Infrastructure.CimCmdlets/CimSessionProxy.cs [hunk 7]
```diff
@@ -875,7 +875,7 @@ internal void WriteOperationStartMessage(string operation, Hashtable parameterLi
                         parameters.Append(',');
                     }
 
-                    parameters.Append(string.Format(CultureInfo.CurrentUICulture, @"'{0}' = {1}", key, parameterList[key]));
+                    parameters.Append(string.Format(CultureInfo.CurrentUICulture, $@"'{key}' = {parameterList[key]}"));
```
Comment (iSazonov) [inline src/Microsoft.Management.Infrastructure.CimCmdlets/CimSessionProxy.cs:878]: We can use StringBuilder.AppendFormat

File: src/Microsoft.Management.Infrastructure.CimCmdlets/Utils.cs [hunk 8]
```diff
@@ -205,12 +205,10 @@ internal static string GetSourceCodeInformation(bool withFileName, int depth)
...
             // else
             // {
-            //    return string.Format(CultureInfo.CurrentUICulture, "{0}:", frame.GetMethod());
+            //    return string.Format(CultureInfo.CurrentUICulture, $"{frame.GetMethod()}:");
             // }
 
-            return string.Format(CultureInfo.CurrentUICulture, "{0}::{1}        ",
-                frame.GetMethod().DeclaringType.Name,
-                frame.GetMethod().Name);
+            return string.Format(CultureInfo.CurrentUICulture, $"{frame.GetMethod().DeclaringType.Name}::{frame.GetMethod().Name}        ");
```
Comment (iSazonov) [inline src/Microsoft.Management.Infrastructure.CimCmdlets/Utils.cs:211]: string.Format ?

File: src/Microsoft.Management.UI.Internal/HelpWindow/HelpParagraphBuilder.cs [hunk 9]
```diff
@@ -369,7 +369,7 @@ private void AddSyntax(bool setting, string sectionTitle)
                     continue;
                 }
 
-                string commandStart = string.Format(CultureInfo.CurrentCulture, "{0} ", commandName);
+                string commandStart = string.Format(CultureInfo.CurrentCulture, $"{commandName} ");
```
Comment (iSazonov) [inline src/Microsoft.Management.UI.Internal/HelpWindow/HelpParagraphBuilder.cs:372]: The same.

File: src/Microsoft.Management.UI.Internal/ManagementList/FilterCore/ValidatingValueBase.cs [hunk 10]
```diff
@@ -218,7 +218,7 @@ internal DataErrorInfoValidationResult EvaluateValidationRules(object value, Sys
                 DataErrorInfoValidationResult result = rule.Validate(value, cultureInfo);
                 if (result == null)
                 {
-                    throw new InvalidOperationException(string.Format(CultureInfo.CurrentCulture, "DataErrorInfoValidationResult not returned by ValidationRule: {0}", rule.ToString()));
+                    throw new InvalidOperationException(string.Format(CultureInfo.CurrentCulture, $"DataErrorInfoValidationResult not returned by ValidationRule: {rule.ToString()}"));
```
Comment (iSazonov) [inline src/Microsoft.Management.UI.Internal/ManagementList/FilterCore/ValidatingValueBase.cs:221]: The same. And we can remove ToString().

File: src/Microsoft.Management.UI.Internal/ManagementList/FilterProviders/SearchTextParser.cs [hunk 11]
```diff
@@ -121,7 +121,7 @@ protected virtual string GetPattern()
                 patterns.Add(rule.Pattern);
             }
 
-            patterns.Add(string.Format(CultureInfo.InvariantCulture, "(?<{0}>){1}", FullTextRuleGroupName, ValuePattern));
+            patterns.Add(string.Format(CultureInfo.InvariantCulture, $"(?<{FullTextRuleGroupName}>){ValuePattern}"));
```
Comment (iSazonov) [inline src/Microsoft.Management.UI.Internal/ManagementList/FilterProviders/SearchTextParser.cs:124]: The same. I stop review. Please replace all Format with Create.

Comment (CarloToso) [reply src/Microsoft.Management.UI.Internal/ManagementList/FilterProviders/SearchTextParser.cs:124]: I'm still working on it, its [WIP]. I'll change it to draft

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImplicitRemotingCommands.cs [hunk 12]
```diff
@@ -2247,13 +2247,13 @@ private string GenerateNewPSSessionOption()
             RunspaceConnectionInfo runspaceConnectionInfo = _remoteRunspaceInfo.Runspace.ConnectionInfo as RunspaceConnectionInfo;
             if (runspaceConnectionInfo != null)
             {
-                result.AppendFormat(null, "-Culture '{0}' ", CodeGeneration.EscapeSingleQuotedStringContent(runspaceConnectionInfo.Culture.ToString()));
-                result.AppendFormat(null, "-UICulture '{0}' ", CodeGeneration.EscapeSingleQuotedStringContent(runspaceConnectionInfo.UICulture.ToString()));
+                result.AppendFormat(null, $"-Culture '{CodeGeneration.EscapeSingleQuotedStringContent(runspaceConnectionInfo.Culture.ToString())}' ");
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImplicitRemotingCommands.cs:2250]: ToString() can be removed in interpolated strings.

**Comment by iSazonov** [comment]
@CarloToso The PR is too large. I suggest to split it by 5-10-15 files.

**Comment by CarloToso** [comment]
@iSazonov You are right this PR is too big, I'll leave it open until until I've finished splitting it, then I'll close it

**Comment by CarloToso** [comment]
PR split in 7 parts

**Pull request closed without merging**